docs/rfcs/027-crash-consistent-layer-map-through-index-part.md

```diff
@@ -0,0 +1,219 @@
+Instead, it treats the remote index part as the authoritative layer map.
+If the local timeline dir contains a layer that is in the remote index part, that's nice, and we'll re-use it if file size (and in the future, check sum) match what's stated in the index part.
+If it doesn't match, we remove the file from the local timeline dir.
+
+After load, **at runtime**, nothing changes compared to what we did before this RFC.
+The procedure for single- and multi-object changes is reproduced here for reference:
+* For any new layers that the change adds:
+  * Write them to a temporary location.
+  * While holding layer map lock:
+    * Move them to the final location.
+    * Insert into layer map.
+* Make the S3 changes.
+  We won't reproduce the remote timeline client method calls here because these are subject to change.
+  Instead we reproduce the sequence of s3 changes that must result for a given single-/multi-object change:
+    * PUT layer files inserted by the change.
+    * PUT an index part that has insertions and deletions of the change.
+    * DELETE the layer files that are deleted by the change.
+      * With the [split-brain protection RFC](https://github.com/neondatabase/neon/pull/4919), the deletions will be written to deletion queue instead of scheduled.
+
+## How This Solves The Problem
+
+If we crash before we've finished the S3 changes, then timeline load will reset layer map to the state that's in the S3 index part.
+The S3 change sequence above is obviously crash-consistent.
+If we crash before the index part PUT, then we leak the inserted layer files to S3.
+If we crash after the index part PUT, we leak the to-be-DELETEd layer files to S3.
+Leaking is fine, it's a pre-existing condition and not addressed in this RFC.
+
+Multi-object changes that previously created and removed files in timeline dir are now atomic because the layer map updates are atomic and crash consistent:
+* atomic layer map update at runtime, currently by using an RwLock in write mode
+* atomic `index_part.json` update in S3, as per guarantee that S3 PUT is atomic
+* local timeline dir state:
+  * irrelevant for layer map content => irrelevant for atomic updates / crash consistency
+  * if we crash after index part PUT, local layer files will be used, so, no on-demand downloads neede for them
+  * if we crash before index part PUT, local layer files will be deleted
+
+## Trade-Offs
+
+`disk_consistent_lsn` can go backwards across restarts if we crash before we've finished the index part PUT.
+Nobody should care about it, because the only thing that matters is `remote_consistent_lsn`.
+Compute certainly doesn't care about `disk_consistent_lsn`.
+
+If we crash before finishing the index part PUT, we have to re-do the work during restart:
+* wal ingest: losing not-yet-uploaded L0s; load on the **safekeepers** + work for pageserver
+* compaction: we lose the entire compaction iteration work; need to re-do it again
```

Comment:
Currently, compaction can be very long running -- is re-doing from start unavoidable, or could we avoid totally re-doing if we modify the compaction code to do some mid-compaction uploads between key ranges?

Reply:
Resumable / incremental compaction is issue https://github.com/neondatabase/neon/issues/4580
I don't know Chi's compaction work well enough to tell whether he solved that as part of his prototype.

---

In this prototype, re-doing the compaction work is unavoidable.

How long running is it in practice, though? Let me see whether we have useful metrics for that.

Reply:
When we tried a 1TiB example the other week, it was hours (although that was including some restarts since it was in staging)

Reply:
So, practically speaking, here are some approximate numbers from prod (we should fix the histogram to not count no-op iterations): https://neonprod.grafana.net/d/d9247703-8c96-44f0-b8d1-cbd4ee17062a/compaction-timings?orgId=1&from=1692706093728&to=1692852222721

Reply:
Was this 1TiB logical size or physical size?

Reply:
logical

Reply:
I think it was reimaging which was taking that long instead of L0 -> L1 compaction. We just do them sequentially.

Reply:
I rewrote the trade-off section, moving fundamental trade-off to the front and tried to add numbers / worst-case estimations where I could.

Please re-read that section and continue discussion in a new comment thread, repeating any remarks from this thread should I have not addressed them.